File: java/parsers/src/main/java/jp/co/teruuu/parser/json/JsonParser.java
package jp.co.teruuu.parser.json;

import jp.co.teruuu.parser.common.Pair2;
import jp.co.teruuu.parser.common.Pair3;
import jp.co.teruuu.parser.common.ParseResult;
import jp.co.teruuu.parser.common.Parser;
import jp.co.teruuu.parser.common.type.Tuple2;
import jp.co.teruuu.parser.common.type.Tuple3;
import jp.co.teruuu.parser.json.type.JValue;

import java.util.List;

public class JsonParser implements JParser<JValue> {
  Parser<List<String>> sp = Parser.string(" ").many();

  @Override
  public ParseResult<JValue> parse(String input) {
    Parser<Tuple2<Tuple3<List<String>, JValue, List<String>>, String>> parser = new Pair2(new Pair3(sp, JParser.object().or(JParser.array()), sp), Parser.EOF());
    ParseResult result = parser.parse(input);
    if(result instanceof ParseResult.Success) {
      Tuple2<Tuple3<List<String>, JValue, List<String>>, String> resultV = ((ParseResult.Success<Tuple2<Tuple3<List<String>, JValue, List<String>>, String>>) result).value;
      return new ParseResult.Success<>(resultV.item1.item2, ((ParseResult.Success<Tuple2<Tuple3<List<String>, JValue, List<String>>, String>>) result).next);
    } else {
      return result;
    }
  }
}

File: scala/parsers/src/main/scala/jp/co/teruuu/parser/json/JsonParser.scala
package jp.co.teruuu.parser.json

import jp.co.teruuu.parser.common._

object JsonParser extends JParser[JValue]{
  lazy val parser = new Pair2(JObjectParser.jor(JArrayParser), Parser.eof)

  override def parse(input: String): ParseResult[JValue] = {
    parser.parse(input) match {
      case result: ParseOk[(JValue, String)] => ParseOk(result.value._1, result.next)
      case result: ParseNg[_] => ParseNg(result.message, result.next)
    }
  }
}
